### PR TITLE
From file

### PR DESCRIPTION
**Pull request by toryjleo**
Take a look and tell me if you want it to work differently

### CODE REVIEW

File: cpp/include/util.h [hunk 1]
```diff
@@ -36,14 +37,43 @@ namespace Nice {
...
-    for (int i = 0; i < num_rows; i++)
-      for (int j = 0; j < num_cols; j++)
-        input_file >> m(i, j);
+    while(!input_file.eof()) {
+      getline(input_file, line);
+      if(line.find_first_not_of(' ') == std::string::npos) {
+        continue;
+      }
+      std::stringstream stream(line);
+      colsinrow = 0;
```
Comment (yiskylee) [inline cpp/include/util.h:93]: cols_in_row

File: cpp/include/util.h [hunk 2]
```diff
@@ -36,14 +37,43 @@ namespace Nice {
...
-Matrix<T> FromFile(const std::string &input_file_path,
-                   int num_rows, int num_cols) {
+Matrix<T> FromFile(const std::string &input_file_path) {
   std::ifstream input_file(input_file_path, std::ifstream::in);
-  Matrix<T> m(num_rows, num_cols);
+  Matrix<T> m;
+  std::string line;
+  std::vector<T> vector;
+  T coef;
+  int cols = 0;
```
Comment (yiskylee) [inline cpp/include/util.h]: num_cols

File: cpp/include/util.h [hunk 3]
```diff
@@ -36,14 +37,43 @@ namespace Nice {
...
+  int rows = 0;
```
Comment (yiskylee) [inline cpp/include/util.h]: num_rows

File: cpp/include/util.h [hunk 4]
```diff
@@ -36,14 +37,43 @@ namespace Nice {
...
 namespace util {
 template<typename T>
-Matrix<T> FromFile(const std::string &input_file_path,
-                   int num_rows, int num_cols) {
+Matrix<T> FromFile(const std::string &input_file_path) {
   std::ifstream input_file(input_file_path, std::ifstream::in);
-  Matrix<T> m(num_rows, num_cols);
+  Matrix<T> m;
+  std::string line;
+  std::vector<T> vector;
```
Comment (yiskylee) [inline cpp/include/util.h]: temp_buffer

File: cpp/include/util.h [hunk 5]
```diff
@@ -36,14 +37,43 @@ namespace Nice {
...
+        std::cerr << "Problem with Matrix in: " + input_file_path + ", exiting...";
+        exit(1);
+      }
+      ++rows;
+    }
+
+    m.resize(rows, cols);
+    for (int i = 0; i < rows; ++i) {
+      for(int j = 0; j < cols; ++j) {
+        m(i, j) = vector.at(i * rows + j);
```
Comment (yiskylee) [inline cpp/include/util.h]: I think Eigen supports converting directly from an std::vector to Eigen::Matrix, you can read about it at: http://stackoverflow.com/questions/17036818/initialise-eigenvector-with-stdvector 
I think in that way we could get rid of the for loop

**Comment by yiskylee** [comment]
This looks great! Couple of thoughs: 1. I don't want to have the previous FromFile(input, num_rows, num_cols) deleted though because it offers users with a faster file reading method if they do know the dimensions of the data. Could you put them back? 2. I made some comments on variable naming and eliminating for loop for std::vector to Eigen copying.

**Comment by toryjleo** [comment]
I made the changes

**Comment by yiskylee** [comment]
There are other comments yet to be addressed. 1. I don't want to have the previous FromFile(input, num_rows, num_cols) deleted though because it offers users with a faster file reading method if they do know the dimensions of the data. Could you put them back? 2. Eliminate for loop for std::vector to Eigen copying (I think Eigen supports converting directly from an std::vector to Eigen::Matrix, you can read about it at: http://stackoverflow.com/questions/17036818/initialise-eigenvector-with-stdvector 
I think in that way we could get rid of the for loop) 3. colsinrow -> cols_in_row

**Comment by yiskylee** [comment]
1. You have also deleted the old tests for the previous FromFile test, could you add them back the same way as how you added the previous FromFile function?
2. Please use ASSERT_NEAR() instead of .isApprox() in your test file. You can read about how to use it at https://github.com/google/googletest/blob/master/googletest/docs/AdvancedGuide.md. Leslie used assert_near as well in her test files.

**Comment by toryjleo** [comment]
The deed has been done, Xiangyu.

**Comment by yiskylee** [comment]
Looks good!

**Pull request merged**